[Clang] Report an error and crash on source location exhaustion in macros

ilya-biryukov opened the pull request:
`createExpansionLocImpl` has an assert that checks if we ran out of source locations. We have observed this happening on real code and in release builds the assertion does not fire and the compiler just keeps running indefinitely without giving any indication that something went wrong.

Diagnose this problem and reliably crash to make sure the problem is easy to detect.

I have also tried:
- returning invalid source locations,
- reporting sloc address space usage on error.

Both caused the compiler to run indefinitely. It would be nice to dig further why that happens, but until then crashing seems like a better alternative.

ilya-biryukov committed:
[SourceManager] Report an error and crash on source location exhaustion in macros

`createExpansionLocImpl` has an assert that checks if we ran out of source
locations. We have observed this happening on real code and in release builds
the assertion does not fire and the compiler just keeps running
indefinitely without giving any indication that something went wrong.

Diagnose this problem and reliably crash to make sure the problem is easy to
detect.

I have also tried:
- returning invalid source locations,
- reporting sloc address space usage on error.

Both caused the compiler to run indefinitely. It would be nice to dig
further why that happens, but until then crashing seems like a better
alternative.

sam-mccall reviewed (commented):
Upgrading this from assert to diag+crash makes sense to me.
Am I missing where the crash happens though?

ilya-biryukov commented on clang/include/clang/Basic/DiagnosticCommonKinds.td:
```diff
@@ -356,6 +356,8 @@ def err_file_modified : Error<
   "file '%0' modified since it was first processed">, DefaultFatal;
 def err_file_too_large : Error<
   "sorry, unsupported: file '%0' is too large for Clang to process">;
+def err_expansions_too_large : Error<
```
I also don't feel strongly there. The reason I introduced a new warning is because this could theoretically happen without any `#include` and I did not want to touch the existing error message.

sam-mccall replied:
SGTM (but the current version is also OK with me)

ilya-biryukov replied:
I have reused the existing diagnostic and updated it to remove a mention of include.

ilya-biryukov committed:
fixup! [SourceManager] Report an error and crash on source location exhaustion in macros

ilya-biryukov commented on clang/lib/Basic/SourceManager.cpp:
```diff
@@ -663,10 +663,13 @@ SourceManager::createExpansionLocImpl(const ExpansionInfo &Info,
+    Diag.Report(Info.getExpansionLocStart(), diag::err_expansions_too_large);
+    Diag.Report(Info.getExpansionLocStart(), diag::remark_sloc_usage);
+    noteSLocAddressSpaceUsage(Diag);
+    return SourceLocation();
```
oops, sorry, I forgot to commit before pushing, changes were in the working index.
Should be good now.

sam-mccall reviewed (approved):
LG but would be great if we could fix the note!

ilya-biryukov commented on clang/lib/Basic/SourceManager.cpp:
```diff
@@ -663,10 +663,15 @@ SourceManager::createExpansionLocImpl(const ExpansionInfo &Info,
+  if (NextLocalOffset + Length + 1 <= NextLocalOffset ||
+      NextLocalOffset + Length + 1 > CurrentLoadedOffset) {
+    Diag.Report(Info.getSpellingLoc(), diag::err_expansions_too_large);
+    // FIXME: call `noteSLocAddressSpaceUsage` to report details to users.
```
I actually tried moving the sloc entry below this check, but it did not help.
So no idea what causes it, and I thought having the change as is is better than the current state, so I decided to sent the PR before figuring this one out.

ilya-biryukov replied:
Another caveat: `Info.getSpellingLoc` can sometimes return one of the expansion locations and this causes the diagnostic printing to run indefinitely too!

So I had to report an error with invalid `SourceLocation()`. The crash stack trace helpfully points at tokens that were being processed, so it does give some context.

sam-mccall replied:
This really sounds like the SourceManager is corrupt at this point, would be good to find out why but this is already an improvement

ilya-biryukov committed:
fixup! [SourceManager] Report an error and crash on source location exhaustion in macros

- Unify the error when running out of slocs in includes and macro
- Use `SourceLocation()` instead of `Info.getSpellingLoc()` to avoid printing
  the location, which also causes Clang to run indefinitely.

ilya-biryukov committed:
fixup! [SourceManager] Report an error and crash on source location exhaustion in macros

- Add a comment about using SourceLocation()

ilya-biryukov committed:
fixup! [SourceManager] Report an error and crash on source location exhaustion in macros

- Update test for sloc overflow

ilya-biryukov commented:
I will land this, but happy to follow up if Aaron or anyone else will have comments.
